### PR TITLE
324 Add GADM

### PR DESCRIPTION
**Pull request by MattBlissett**
This adds interpretation through the existing geocode KVS, and handles the HDFS table (NB needs branch version of occurrence).

@fmendezh / @muttcg, advice requested.

We have eight new fields:
- level0gid, level0name
- level1gid, level1name
- level2gid, level2name
- level3gid, level3name

We'll get this sort of data:

```
 type  │    gid      │      source      │ name  
───────┼─────────────┼──────────────────┼────────
 GADM0 │ JPN         │ http://gadm.org/ │ Japan   
 GADM1 │ JPN.26_1    │ http://gadm.org/ │ Nagano  
 GADM2 │ JPN.26.40_1 │ http://gadm.org/ │ Nagawa 
```

 As Tim mentioned at https://github.com/gbif/pipelines/issues/324#issuecomment-658253127

> Please consider that we anticipate GADM will be used a lot (e.g. create a choropleth map with an aggregation code, metrics for country pages aggregated by county) which may influence the ES structure necessary.

Can you think about this? I don't know enough about ES to decide on the appropriate structure.

It's also just occurred to me that a user might search "長野県" or "長和町". Do we need this too?

### CODE REVIEW

File: sdks/parsers/src/main/java/org/gbif/pipelines/parsers/parsers/location/parser/LocationMatcher.java [hunk 1]
```diff
@@ -131,6 +142,24 @@ public LocationMatcher additionalTransform(UnaryOperator<LatLng> transformation)
     return Optional.empty();
   }
 
+  private Optional<GadmFeatures> getGadmFromCoordinates(LatLng latLng) {
+    if (latLng.isValid()) {
+      GeocodeResponse geocodeResponse = null;
+      try {
+        geocodeResponse = geocodeKvStore.get(latLng);
+      } catch (NoSuchElementException | NullPointerException ex) {
```
Comment (fmendezh) [inline sdks/parsers/src/main/java/org/gbif/pipelines/parsers/parsers/location/parser/LocationMatcher.java:150]: catching those exceptions here looks like there's bug somewhere else

File: sdks/parsers/src/main/java/org/gbif/pipelines/parsers/parsers/location/parser/LocationParser.java [hunk 2]
```diff
@@ -99,6 +100,22 @@
...
+    Objects.requireNonNull(lr, "LocationRecord is required");
+    Objects.requireNonNull(kvStore, "GeocodeService kvStore is required");
+
+    // Take parsed values
+    LatLng location = new LatLng(lr.getDecimalLatitude(), lr.getDecimalLongitude());
+
+    // Use these to retrieve the GADM areas.
+    Optional<GadmFeatures> match =
+        LocationMatcher.create(location, null, kvStore)
+            .applyGadm();
```
Comment (muttcg) [inline sdks/parsers/src/main/java/org/gbif/pipelines/parsers/parsers/location/parser/LocationParser.java:113]: Could be replaced to 
```
return LocationMatcher.create(location, null, kvStore)
            .applyGadm();
```
or added debug logger

File: pipelines/ingest-transforms/src/test/java/org/gbif/pipelines/transforms/core/LocationTransformTest.java [hunk 3]
```diff
@@ -53,9 +54,59 @@ public void processElement(ProcessContext context) {
   public final transient TestPipeline p = TestPipeline.create();
 
   private static GeocodeResponse toGeocodeResponse(Country country) {
-    Location location = new Location();
-    location.setIsoCountryCode2Digit(country.getIso2LetterCode());
-    return new GeocodeResponse(Collections.singletonList(location));
+    List<Location> locations = new ArrayList();
```
Comment (muttcg) [inline pipelines/ingest-transforms/src/test/java/org/gbif/pipelines/transforms/core/LocationTransformTest.java:57]: Raw usage
`new ArrayList<>();`

File: sdks/parsers/src/main/java/org/gbif/pipelines/parsers/parsers/location/cache/GeocodeBitmapCache.java [hunk 4]
```diff
@@ -24,8 +23,7 @@
   // World map image lookup
   private final BufferedImage img;
   private static final int BORDER = 0x000000;
-  private static final int EEZ = 0x888888;
-  private static final int INTERNATIONAL_WATER = 0xFFFFFF;
+  private final static int NOTHING = 0xFFFFFF;
```
Comment (muttcg) [inline sdks/parsers/src/main/java/org/gbif/pipelines/parsers/parsers/location/cache/GeocodeBitmapCache.java:26]: Java specs recommend "static final" order

File: sdks/parsers/src/main/java/org/gbif/pipelines/parsers/parsers/location/parser/GadmFeatures.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            level2Gid = l.getId();
+            level2Name = l.getName();
+          }
+          return;
+        case "GADM3":
+          if (level3Gid == null) {
+            level3Gid = l.getId();
+            level3Name = l.getName();
+          }
+          return;
```
[the remark this answers is not in the document]
Comment (MattBlissett) [reply sdks/parsers/src/main/java/org/gbif/pipelines/parsers/parsers/location/parser/GadmFeatures.java:45]: I won't remove 45, even here it's not great to use a fall-through.  [Google Code style](https://google.github.io/styleguide/javaguide.html#s4.8.4-switch) requires the `default:`, the `return` in it doesn't really matter.

I left it as there are other values coming through, like `EEZ` and `Political`.

(When compiled, I doubt any of these variations make any difference.)

File: sdks/parsers/src/main/java/org/gbif/pipelines/parsers/parsers/location/parser/LocationMatcher.java (hunk 1 above)
Comment (MattBlissett) [reply sdks/parsers/src/main/java/org/gbif/pipelines/parsers/parsers/location/parser/LocationMatcher.java:150]: It's not required. I'd copied the call from the method above (the original geocoding) without checking.

Removing here and above.

File: pipelines/ingest-gbif/src/main/resources/elasticsearch/es-occurrence-schema.json [hunk 6]
```diff
@@ -118,6 +118,20 @@
...
+        "type": "object",
+        "properties": {
+          "gadmLevel0Gid": {"type": "keyword"},
+          "gadmLevel1Gid": {"type": "keyword"},
+          "gadmLevel2Gid": {"type": "keyword"},
+          "gadmLevel3Gid": {"type": "keyword"},
+          "gadmLevel0Name": {"type": "text"},
+          "gadmLevel1Name": {"type": "text"},
+          "gadmLevel2Name": {"type": "text"},
+          "gadmLevel3Name": {"type": "text"}
```
[the remark this answers is not in the document]
Comment (MattBlissett) [reply pipelines/ingest-gbif/src/main/resources/elasticsearch/es-occurrence-schema.json:131]: Should it be `gadm.level0Gid` then?

[the remark this answers is not in the document]
Comment (muttcg) [reply pipelines/ingest-gbif/src/main/resources/elasticsearch/es-occurrence-schema.json:131]: Yes, I think it should be:
```
gadm
- level0Gid
- level0Name
- ..ect
```

**Pull request merged**